go/gilded-rose.go

```diff
@@ -1,57 +1,155 @@
 package main
 
+import (
+	"fmt"
+)
+
 type Item struct {
 	name            string
 	sellIn, quality int
 }
 
-func UpdateQuality(items []*Item) {
-	for i := 0; i < len(items); i++ {
+func (item *Item) Name() string {
+	return item.name
+}
+
+func (item *Item) addQuality(val int) error {
+	if val < 0 {
+		return fmt.Errorf("addQuality: val is negative, %v", val)
+	}
+
+	item.quality = item.quality + val
+	return nil
+}
+
+func (item *Item) subQuality(val int) error {
+	if val < 0 {
+		return fmt.Errorf("subQuality: val is negative, %v", val)
+	}
+
+	item.quality = item.quality - val
+	return nil
+}
+
+func (item *Item) incQuality() error {
+	return item.addQuality(1)
+}
+
+func (item *Item) decQuality() error {
+	return item.subQuality(1)
+}
+
+func (item *Item) resetQuality() error {
+	item.quality = 0
+	return nil
+}
+
+func (item *Item) Quality() int {
+	return item.quality
+}
+
+func (item *Item) subSellIn(val int) error {
+	if val < 0 {
+		return fmt.Errorf("subSellIn: val is negative, %v", val)
+	}
 
-		if items[i].name != "Aged Brie" && items[i].name != "Backstage passes to a TAFKAL80ETC concert" {
-			if items[i].quality > 0 {
-				if items[i].name != "Sulfuras, Hand of Ragnaros" {
-					items[i].quality = items[i].quality - 1
-				}
-			}
-		} else {
-			if items[i].quality < 50 {
-				items[i].quality = items[i].quality + 1
-				if items[i].name == "Backstage passes to a TAFKAL80ETC concert" {
-					if items[i].sellIn < 11 {
-						if items[i].quality < 50 {
-							items[i].quality = items[i].quality + 1
-						}
-					}
-					if items[i].sellIn < 6 {
-						if items[i].quality < 50 {
-							items[i].quality = items[i].quality + 1
-						}
-					}
-				}
-			}
+	item.sellIn = item.sellIn - val
+	return nil
+}
+
+func (item *Item) decSellIn() error {
+	return item.subSellIn(1)
+}
+
+func (item *Item) SellIn() int {
+	return item.sellIn
+}
+
+func (item *Item) defaultCase() error {
+	if item.Quality() > 0 {
+		item.decQuality()
+		if item.SellIn() <= 0 && item.Quality() > 0 {
+			item.decQuality()
 		}
+	}
+	item.decSellIn()
 
-		if items[i].name != "Sulfuras, Hand of Ragnaros" {
-			items[i].sellIn = items[i].sellIn - 1
+	return nil
+}
+
+func (item *Item) sulfurasCase() error {
+	return nil
+}
+
+func (item *Item) backstageCase() error {
+	item.decSellIn()
+	if item.SellIn() < 0 {
+		item.resetQuality()
+		return nil
+	}
+
+	if item.Quality() < 50 {
+		item.incQuality()
+		if item.SellIn() < 10 && item.Quality() < 50 {
+			item.incQuality()
+		}
+		if item.SellIn() < 5 && item.Quality() < 50 {
+			item.incQuality()
 		}
+	}
```

Comment:
this logic can easily get written in GO in a more readable manner,
first if you `check item.Quality() < 50`, then in the next `if` statements you don't need the `&&`.
in any case think about this:
```suggestion
if item.Quality() < 50 {
		switch {
		case item.SellIn() < 10:
			item.addQuality(2)
		case item.SellIn() < 5:
			item.addQuality(3)
		default:
			item.addQuality(1)
		}
	}
```